File: tests/orig_parity/08_nhefs_ch14_gestimation.py
"""StatsPAI original-data parity (Python side) -- Module 08.

Reproduces Hernán & Robins, *Causal Inference: What If*, **Chapter 14**
(G-estimation of a structural nested mean model) on the REAL NHEFS data
bundled in ``sp.datasets.nhefs()``.  Estimand: ``psi``, the additive
(rank-preserving) one-parameter SNMM coefficient for the effect of
quitting smoking (``qsmk``) on 10-year weight change (``wt82_71``, kg).

Compares against:
  (a) the book's published number -- Program 14.2: ``psi`` ~ 3.4 kg,
      95% CI (2.5, 4.5);
  (b) the R-side base-R reference run on the same CSV bytes
      (``08_nhefs_ch14_gestimation.R``), which implements BOTH the book's
      exact *logistic* g-estimation (fit ``qsmk ~ H(psi) + L``, solve for
      the ``psi`` at which the coefficient on ``H(psi)`` is zero) and the
      *linear* moment-condition g-estimator that StatsPAI uses.

CONVENTION NOTE.  Since 1.29 ``sp.g_estimation`` defaults to
``propensity_model='logit'``: it solves ``sum (A - A_hat)(Y - psi A) = 0``
jointly with the treatment-free OLS equations, ``A_hat`` a logistic fit on
``L``. Because the logit score makes ``A - A_hat`` orthogonal to ``L``, this
is exactly Program 14.2's closed form -- the ``psi`` at which the
coefficient on ``H(psi) = Y - psi*A`` vanishes in ``qsmk ~ H(psi) + L``
(3.4611 on NHEFS, the R side's ``snmm_psi``). Before 1.29 StatsPAI used the
*linear* moment condition (``psi = cov(Y_res, A_res) / var(A_res)`` with OLS
residuals; 3.4626), which ``propensity_model='linear'`` still reproduces and
the R side still carries as ``snmm_psi_linear``.
"""

from __future__ import annotations

import numpy as np
import statspai as sp

from _common import OrigRecord, write_results
from _nhefs import book_design, dump_csv

MODULE = "08_nhefs_ch14_gestimation"


def main() -> None:
    df = dump_csv(MODULE)  # n=1566 complete case; identical bytes for R
    n = len(df)

    # StatsPAI additive-SNMM g-estimate of psi via sp.g_estimation
    # (single stage: the point-treatment qsmk effect on wt82_71).
    dd, covs = book_design(df)
    res = sp.g_estimation(
        dd,
        y="wt82_71",
        treatments=["qsmk"],
        covariates_by_stage=[covs],
        n_bootstrap=500,
        random_state=42,
    )
    psi = float(res.estimate)
    se = float(res.se)
    ci = (float(res.ci[0]), float(res.ci[1]))

    # Independent in-script reproduction of the linear moment condition
    # (StatsPAI's pre-1.29 default, propensity_model='linear') -- the value
    # the R "linear g-estimation" anchor must match.
    X = dd[covs].values.astype(float)
    A = df["qsmk"].values.astype(float)
    Y = df["wt82_71"].values.astype(float)
    Xa = np.column_stack([np.ones(n), X])
    beta_y, *_ = np.linalg.lstsq(Xa, Y, rcond=None)
    beta_a, *_ = np.linalg.lstsq(Xa, A, rcond=None)
    Yres = Y - Xa @ beta_y
    Ares = A - Xa @ beta_a
    psi_linear = float(np.sum(Yres * Ares) / np.sum(Ares**2))

    rows = [
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="snmm_psi",
            estimate=psi,
            se=se,
            n=n,
            published=3.4,
            citation="Hernán-Robins, What If Program 14.2 (G-estimation SNMM psi)",
            extra={
                "ci": [ci[0], ci[1]],
                "pvalue": float(res.pvalue),
                "method": res.method,
            },
        ),
        OrigRecord(
            module=MODULE,
            side="py",
            statistic="snmm_psi_linear_check",
            estimate=psi_linear,
            se=None,
            n=n,
            published=3.4,
            citation="Linear moment-condition g-estimate (propensity_model='linear')",
        ),
    ]

    write_results(
        MODULE,
        "py",
        rows,
        extra={
            "data_source": "sp.datasets.nhefs (NHEFS / What If)",
            "n_obs": n,
            "estimand": "additive SNMM psi (qsmk -> wt82_71)",
            "ci_snmm": [ci[0], ci[1]],
            "published_psi": 3.4,
            "published_snmm_ci": [2.5, 4.5],
            "statspai_default_propensity_model": "logit",
        },
    )
    print(
        f"[{MODULE}] g-est psi={psi:.4f} 95% CI "
        f"({ci[0]:.2f},{ci[1]:.2f})  linear_check={psi_linear:.4f}  "
        f"(book 3.4 [2.5,4.5])"
    )


if __name__ == "__main__":
    main()

File: tests/orig_parity/08_nhefs_ch14_gestimation.R
# Original-data parity for Hernán & Robins, *Causal Inference: What If*,
# Chapter 14 (G-estimation of a structural nested mean model).
#
#   Source data : sp.datasets.nhefs() complete case (n=1566), dumped by
#                 the Python side to data/08_nhefs_ch14_gestimation.csv so
#                 both languages read identical bytes.
#   Published   : Program 14.2 -- additive SNMM psi ~ 3.4 kg,
#                 95% CI (2.5, 4.5).
#
# This script carries TWO base-R references:
#
#   (1) snmm_psi_logistic -- the book's EXACT Program 14.2 procedure.
#       For a candidate psi form H(psi) = wt82_71 - psi*qsmk, fit the
#       logistic propensity model qsmk ~ H(psi) + L over the canonical
#       confounders, and find the psi at which the coefficient on H(psi)
#       is exactly zero (root-find).  The 95% CI is read off where that
#       coefficient's Wald z-statistic crosses +/-1.96 (book's approach).
#
#   (2) snmm_psi_linear -- the linear moment-condition g-estimator that
#       StatsPAI's sp.g_estimation implements: residualise wt82_71 and
#       qsmk on L (OLS) and set psi = cov(Yres, Ares)/var(Ares).  This is
#       the tight (machine-precision) cross-language anchor for the
#       StatsPAI point estimate.
#
# Writes results/08_nhefs_ch14_gestimation_R.json.

.script_dir <- (function() {
  args <- commandArgs(trailingOnly = FALSE)
  m <- grep("^--file=", args, value = TRUE)
  if (length(m) > 0) dirname(normalizePath(sub("^--file=", "", m[1])))
  else getwd()
})()
HERE <- .script_dir
DATA_DIR <- file.path(HERE, "data")
RESULTS_DIR <- file.path(HERE, "results")
dir.create(RESULTS_DIR, showWarnings = FALSE, recursive = TRUE)

suppressPackageStartupMessages({
  library(jsonlite)
})

MODULE <- "08_nhefs_ch14_gestimation"
df <- read.csv(file.path(DATA_DIR, paste0(MODULE, ".csv")))
n <- nrow(df)

A <- df$qsmk
Y <- df$wt82_71

# Canonical confounder set (book Programs 12.x-15.x): categoricals as
# factors, continuous with quadratics -- identical to _nhefs.book_design.
conf <- paste0(
  "factor(sex)+factor(race)+factor(education)+factor(exercise)+",
  "factor(active)+age+I(age^2)+smokeintensity+I(smokeintensity^2)+",
  "smokeyrs+I(smokeyrs^2)+wt71+I(wt71^2)"
)

# ----- (1) Book Program 14.2: logistic g-estimation -----------------------
# For candidate psi, return the coefficient on H(psi) (and its Wald z) in
# the logistic model qsmk ~ H(psi) + L.
fit_H <- function(psi) {
  df$Hpsi <- Y - psi * A
  fml <- as.formula(paste("qsmk ~ Hpsi +", conf))
  m <- glm(fml, data = df, family = binomial())
  s <- summary(m)$coefficients
  c(coef = unname(s["Hpsi", "Estimate"]),
    z = unname(s["Hpsi", "Estimate"] / s["Hpsi", "Std. Error"]))
}

coef_on_H <- function(psi) fit_H(psi)["coef"]
wald_z    <- function(psi) fit_H(psi)["z"]

psi_logistic <- uniroot(coef_on_H, lower = 2.0, upper = 5.0,
                        tol = 1e-8)$root
# CI: psi values where the Wald z of the H(psi) coefficient hits +/-1.96.
ci_lo <- uniroot(function(p) wald_z(p) - 1.96,
                 lower = 0.0, upper = psi_logistic, tol = 1e-6)$root
ci_hi <- uniroot(function(p) wald_z(p) + 1.96,
                 lower = psi_logistic, upper = 7.0, tol = 1e-6)$root

# ----- (2) Linear moment-condition g-estimator (StatsPAI algorithm) -------
des <- as.formula(paste("~", conf))
X <- model.matrix(des, data = df)          # includes intercept
Yres <- residuals(lm.fit(X, Y))
Ares <- residuals(lm.fit(X, A))
psi_linear <- sum(Yres * Ares) / sum(Ares^2)

rows <- list(
  list(module = MODULE, side = "R", statistic = "snmm_psi",
       estimate = psi_logistic, se = NULL, n = n, published = 3.4,
       citation = "Hernán-Robins, What If Program 14.2 (logistic g-estimation)",
       extra = list(ci = c(ci_lo, ci_hi))),
  list(module = MODULE, side = "R", statistic = "snmm_psi_linear",
       estimate = psi_linear, se = NULL, n = n, published = 3.4,
       citation = "Linear moment-condition g-estimate (StatsPAI algorithm)",
       extra = list())
)
payload <- list(
  module = MODULE, side = "R", rows = rows,
  extra = list(engine = "base R glm root-find (Program 14.2) + lm.fit linear g-est",
               estimand = "additive SNMM psi (qsmk -> wt82_71)",
               ci_snmm = c(ci_lo, ci_hi),
               published_psi = 3.4,
               published_snmm_ci = c(2.5, 4.5))
)
writeLines(toJSON(payload, auto_unbox = TRUE, null = "null", digits = 10),
           file.path(RESULTS_DIR, paste0(MODULE, "_R.json")))
cat(sprintf(
  "[%s] logistic psi=%.4f 95%% CI (%.2f,%.2f)  linear psi=%.4f  (book 3.4 [2.5,4.5])\n",
  MODULE, psi_logistic, ci_lo, ci_hi, psi_linear))
